client/allocrunner/alloc_runner_test.go

```diff
@@ -557,3 +561,61 @@ func TestAllocRunner_DeploymentHealth_Unhealthy_Checks(t *testing.T) {
 	require.Equal(t, allochealth.AllocHealthEventSource, last.Type)
 	require.Contains(t, last.Message, "by deadline")
 }
+
+// TestAllocRunner_Destroy asserts that Destroy kills and cleans up a running
+// alloc.
+func TestAllocRunner_Destroy(t *testing.T) {
+	t.Parallel()
+
+	// Ensure task takes some time
+	alloc := mock.BatchAlloc()
+	task := alloc.Job.TaskGroups[0].Tasks[0]
+	task.Config["run_for"] = "10s"
+
+	conf, cleanup := testAllocRunnerConfig(t, alloc)
+	defer cleanup()
+
+	// Use a MemDB to assert alloc state gets cleaned up
+	conf.StateDB = state.NewMemDB()
+
+	ar, err := NewAllocRunner(conf)
+	require.NoError(t, err)
+	go ar.Run()
+
+	// Wait for alloc to be running
+	testutil.WaitForResult(func() (bool, error) {
+		state := ar.AllocState()
+
+		return state.ClientStatus == structs.AllocClientStatusRunning,
+			fmt.Errorf("got client status %v; want running", state.ClientStatus)
+	}, func(err error) {
+		require.NoError(t, err)
+	})
+
+	// Now destroy
+	ar.Destroy()
+
+	select {
+	case <-ar.DestroyCh():
+		// Destroyed properly!
+	case <-time.After(10 * time.Second):
+		require.Fail(t, "timed out waiting for alloc to be destroyed")
+	}
+
+	// Assert alloc is dead
+	state := ar.AllocState()
+	require.Equal(t, structs.AllocClientStatusComplete, state.ClientStatus)
+
+	// Check the state was cleaned
+	allocs, badAllocs, err := conf.StateDB.GetAllAllocations()
```

Comment:
Slightly pedantic and totally ignorable but we may also want to assert that we have 1 allocation before we actually do the cleanup.

Reply:
Good call! My assertion wasn't correct!